draft-ietf-mls-architecture.md

```diff
@@ -316,7 +316,7 @@ The Authentication Service (AS) has to provide three functionalities:
    is valid with respect to a reference identifier
 
 3. Enable a group member to verify that a credential represents the same
```

Comment:
I just realized that this is too restrictive. This can't be a requirement because some infras explicitly want to achieve the opposite. So this must be taken out of the "has to" and rephrased as... "Additionally, an AS can provide..."

Reply:
> I just realized that this is too restrictive. This can't be a requirement because some infras explicitly want to achieve the opposite. So this must be taken out of the "has to" and rephrased as... "Additionally, an AS can provide..."

I don't see the problem here. The requirement is that if two credentials represent the same instance, then other members can determine that. It does not require a system to have multiple instances. In fact if a system only has one instance then by this logic, the other members need to be able to determine that the one instance matches itself.

Reply:
The notion of "instance" doesn't exist in MLS, so anyway this needs to change. Also it seems application specific so it should be removed entirely. Reading between the lines, I think you might mean something like:

"Additionally, in cases where applications require to identify relationships between credentials of the same user, the AS has to provide this functionality". Is this correct ?

Note, however, that it is not what is currently written. The current sentence makes it general to all credential, which we specifically don't want in MLS as some Applications might explicitly want to avoid that.

Reply:
The concern this item was addressing was making sure it's possible to implement the requirements in [Section 13.2](https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#name-proposal-list-validation) of mls-protocol where a proposal list is invalid if:

```
* It contains multiple Add proposals that contain KeyPackages that
  represent the same client according to the application
* It contains an Add proposal with a KeyPackage that represents a
  client already in the group according to the application
```

How is the application supposed to implement these checks if there is not an Authentication Service API for it?

Reply:
I think the phrasing in the protocol doc might have to be changed because these are unnecessary restrictions on the protocol. As long as the key packages are different it is fine to add and remove the same user. It should be the responsibility of the application to signal this restriction through an extension. @bifurcation I think I can see what the intent was, is it to avoid issues on remove when double joining with different key package ?

Reply:
@raphaelrobert @kkohbrok interested in your opinion too. My concern is about mandating the AS to say which keys are related.

Reply:
> Additionally, in cases where applications require to identify relationships between credentials of the same user, the AS has to provide this functionality". Is this correct ?
No. Not user. Instance of user.  

> I think the phrasing in the protocol doc might have to be changed because these are unnecessary restrictions on the protocol. As long as the key packages are different it is fine to add and remove the same user. It should be the responsibility of the application to signal this restriction through an extension. @bifurcation I think I can see what the intent was, is it to avoid issues on remove when double joining with different key package ?
No.

Benjamin, most IM systems support multiple clients/devices/instances for the same (often human) user. These are different keys, different identifiers, and they are treated by the application protocol, and by DoubleRatchet, differently. Applying MLS to these systems, each client/instance has it's own credential. Other MLS clients need to be able to tell when one client/instance wants to replace itself, and need to detect when the same client/instance tries to join twice (ex: because of a programming error, because it crashed at the wrong moment and lost state, or because it got very unlucky missing certain combinations of welcome or commit messages).

I someone wants to use the same credential on multiple devices running MLS, my first reaction would be that this is a very bad idea from a security perspective and out of scope of the MLS architecture document, but they are free to write an Internet draft describing their design to explain when/how/why it would be safe in their environment.

Reply:
I'm not entirely sure what are are discussing here: Do we want the AS to be able to tell

* if two individual clients (via their KeyPackages) belong to the same user, or 
* if two individual KeyPackages belong to the same client? 

I understand the quote from Section 13.2 to require the latter, although the change to the doc says something else.

My suggestion would thus be
```
3. Enable a group member to verify that a credential represents the same client as another credential
```
which would enable performing the checks mentioned by @Bren2010 and also, if I understand them correctly, address the concerns brought up by @rohan-wire.

Reply:
From reading the discussion and the protocol doc again, I think this phrasing is fine, thank you. I think it might have been the intent but the phrasing was confusing. @Bren2010 Does that work for you?

Reply:
Look like we are converging, but fwiw I think the AS should indeed be the component that links devices to accounts. For clients, it is important to track users/clients across groups as noted in #107 and introduced in #109. Note that I don't see any implication regarding privacy, because a well-designed AS would not leak information beyond what clients need to know.

Reply:
Wording changed to @kkohbrok's